File: src/experiments/exp_table_6.py
import time
import os
import numpy as np
import pandas as pd
from scipy.io import savemat

from ..utility import *
from ..methods.local_search import *

problems = ["P123", "P1Sym"]
problem = problems[1]

column_names = ["m", "n", "r", "d", "||A||_0", "||A^+||_0", "||A^+||_1", "||H||_0" , "||H||_1", "time"]

exp = "table_6"
time_limit = 86400

matrices_folder = f"./instances/{exp}"
results_folder = f"./results/{exp}"
solutions_folder = f"./solutions/{exp}"

methods = {
    "P123": "LSLAFI_Det_P3",
    "P1" : "LSLAFI_Det",
    "P1Sym" : "LSLAFI_Det_Symmetric"
}
method = methods[problem]

df = pd.DataFrame(columns=column_names)

instances = get_instances(matrices_folder)

min_unsolvable_m = np.inf

for i in range(1, 6):
    A_norm_0 = 0
    AMP_norm_0 = 0
    AMP_norm_1 = 0
    H_norm_0 = 0
    H_norm_1 = 0
    runtime = 0

    instance = f"{matrices_folder}/Maragal_{i}"
    print(f"Solving for matrix: {instance}")
    A = read_maragal_instance(instance)
    m, n = A.shape
    r = matrix_rank(A)
    d = matrix_vec_0_norm(A) / (m * n)
    d = int(d * 100)
    hatA_flag = 1
    if m < min_unsolvable_m:

        if hatA_flag:
            A = A.T @ A
            r = matrix_rank(A)

        A_MP = np.linalg.pinv(A)

        start_time = time.time()
        H = ls_maragal(exp, solutions_folder, problem, instance, method, hatA_flag, time_limit)
        end_time = time.time()

        if not isinstance(H, int):
            A_norm_0 = matrix_vec_0_norm(A)
            AMP_norm_0 = matrix_vec_0_norm(A_MP)
            AMP_norm_1 = matrix_vec_1_norm(A_MP)
            H_norm_0 = matrix_vec_0_norm(H)
            H_norm_1 = matrix_vec_1_norm(H)
            runtime = end_time - start_time
        else:
            print("TimeLimit: Could not solve instance within the time limit.")
            min_unsolvable_m = m
    instance_results = {
        "m": m,
        "n": n,
        "r": r,
        "d": d,
        "||A||_0": A_norm_0,
        "||A^+||_0": AMP_norm_0,
        "||A^+||_1": AMP_norm_1,
        "||H||_0": H_norm_0,
        "||H||_1": H_norm_1,
        "time": runtime
    }

    df.loc[len(df)] = instance_results

results_filename = f"results_{problem}_LS.csv"
results_filepath = f"{results_folder}/{results_filename}"
df.to_csv(results_filepath, index=False)

File: src/experiments/exp_table_6.m
% exp_table_6.m

clc;
clear;

current_folder = fileparts(mfilename('fullpath'));
methods_path = fullfile(current_folder, '..', 'methods');
addpath(methods_path);

problems = ["P123", "P1Sym"];
problem = problems(2);

column_names = ["m", "n", "r", "d", "||A||_0", "||A^+||_0", "||A^+||_1", "||H||_0", "||H||_1", "time"];

exp = "table_6";
time_limit = 86400;
tol = 1e-5;

matrices_folder = "../../instances/" + exp;
results_folder = "../../results/" + exp;
solutions_folder = "../../solutions/" + exp + "/LS/";

methods = containers.Map;
methods("P123") = "LSLAFI_Det_P3";
methods("P1") = "LSLAFI_Det";
methods("P1Sym") = "LSLAFI_Det_Symmetric";

method = methods(problem);

% Initialize results table
df = table('Size',[0 length(column_names)], ...
           'VariableTypes', repmat("double",1,length(column_names)), ...
           'VariableNames', column_names);

min_unsolvable_m = inf;

for i = 1:5

    A_norm_0 = 0;
    AMP_norm_0 = 0;
    AMP_norm_1 = 0;
    H_norm_0 = 0;
    H_norm_1 = 0;
    runtime = 0;

    instance = matrices_folder + "/Maragal_" + i;
    fprintf("Solving for matrix: %s\n", instance);

    data = load(instance);
    A = data.Problem.A;

    if issparse(A)
        A = full(A);
    end

    [m, n] = size(A);
    r = rank(A, tol);

    d = sum(abs(A(:)) > tol) / (m * n);

    hatA_flag = 1;

    if m < min_unsolvable_m

        if hatA_flag == 1
            A = A' * A;
            r = rank(A, tol);
        end

        A_MP = pinv(A);

        savePath = solutions_folder + "/H_Maragal_" + i + ".mat";

        % ===== TIME MEASUREMENT =====
        t_start = tic;
        success = ls_maragal(instance, r, m, n, method, savePath, hatA_flag, time_limit);
        runtime = toc(t_start);
        % ============================

        if success ~= false

            sol = load(savePath);
            H = sol.H;

            A_norm_0 = sum(abs(A(:)) > tol);
            AMP_norm_0 = sum(abs(A_MP(:)) > tol);
            AMP_norm_1 = sum(sum(abs(A_MP)));
            H_norm_0 = sum(abs(H(:)) > tol);
            H_norm_1 = sum(sum(abs(H)));

        else
            fprintf("TimeLimit: Could not solve instance within the time limit.\n");
            min_unsolvable_m = m;
        end
    end

    % Store results
    new_row = {m, n, r, d, A_norm_0, AMP_norm_0, AMP_norm_1, H_norm_0, H_norm_1, runtime};
    df = [df; new_row];

end

% Save CSV
results_filename = "results_" + problem + "_LS.csv";
results_filepath = results_folder + "/" + results_filename;

writetable(df, results_filepath);

fprintf("Results saved to %s\n", results_filepath);